cranelift/codegen/src/isa/aarch64/lower_inst.rs

```diff
@@ -191,21 +244,71 @@ pub(crate) fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
         }
 
         Opcode::Imul => {
-            let rd = get_output_reg(ctx, outputs[0]).only_reg().unwrap();
-            let rn = put_input_in_reg(ctx, inputs[0], NarrowValueMode::None);
-            let rm = put_input_in_reg(ctx, inputs[1], NarrowValueMode::None);
+            let lhs = put_input_in_regs(ctx, inputs[0]);
+            let rhs = put_input_in_regs(ctx, inputs[1]);
+            let dst = get_output_reg(ctx, outputs[0]);
+
+            let rd = dst.regs()[0];
+            let rn = lhs.regs()[0];
+            let rm = rhs.regs()[0];
+
             let ty = ty.unwrap();
-            if !ty.is_vector() {
-                let alu_op = choose_32_64(ty, ALUOp3::MAdd32, ALUOp3::MAdd64);
-                ctx.emit(Inst::AluRRRR {
-                    alu_op,
-                    rd,
-                    rn,
-                    rm,
-                    ra: zero_reg(),
-                });
-            } else {
-                if ty == I64X2 {
+            match ty {
+                I128 => {
+                    assert_eq!(lhs.len(), 2);
+                    assert_eq!(rhs.len(), 2);
+                    assert_eq!(dst.len(), 2);
+
+                    // 128bit mul formula:
+                    //   dst_lo = lhs_lo * rhs_lo
+                    //   dst_hi = umulhi(lhs_lo, rhs_lo) + (lhs_lo * rhs_hi) + (lhs_hi * rhs_lo)
+                    //
+                    // We can convert the above formula into the following
+                    // umulh   tmp, lhs_lo, rhs_lo
+                    // madd    tmp, lhs_hi, rhs_lo, tmp
+                    // madd    dst_hi, lhs_lo, rhs_hi, tmp
+                    // mul     dst_lo, lhs_lo, rhs_lo
```

Comment:
This should be the second instruction in the sequence, since it can execute in parallel with `UMULH`.

Reply:
It looks like when we do this, we do get a few extra mov's in the emitted code.

When placing the `mul` as the first or second instruction in the sequence we get the following code:
```
umulh x5, x0, x2
madd x4, x0, x2, xzr
madd x0, x0, x3, x5
madd x0, x1, x2, x0
mov x1, x0
mov x0, x4
```

Which makes sense, since we need x0 (lhs_lo) for the 2nd `madd`. I don't know if this is preferred to the 4 instruction codegen? It may be faster, but is longer?

Adding it after the second `madd` (as the third instruction in the sequence) gets us back to a 4 instruction result. But I'm not sure if `madd` executes in parallel with `mul`, I had a quick look at the Software Optimization guide for a few different cores and it doesn't look like it.

```
umulh x4, x0, x2
madd x3, x0, x3, x4
madd x0, x0, x2, xzr
madd x1, x1, x2, x3
```

Let me know which one you prefer

Reply:
IMHO we should prefer code that keeps the regalloc happy and reduces moves (those can have a measurable impact as code-fetch bandwidth can be a limiting factor in some cases) to small reorderings, at least on modern out-of-order cores that can reason about dependencies and issue the `mul` as soon as possible. I'll defer to @akirilov-arm as to whether it makes sense to try harder here for older in-order microarchitectures though!

Reply:
Are you testing with a simple function that just does the multiplication and returns the product? If yes, I wouldn't be too bothered by the additional moves because that's an overly simplified test case (and yes, the longer version might still be faster on an in-order core like the Cortex-A55 that I have in mind with my comments), but you can also try another experiment - reverse the order of the last 2 instructions and see what you get.

Reply:
Yeah, these are all for the function that is in tests:
```
function %mul_i128(i128, i128) -> i128 {
block0(v0: i128, v1: i128):
    v2 = imul v0, v1
    return v2
}
```

> but you can also try another experiment - reverse the order of the last 2 instructions and see what you get.

Do you mean the last `mul` and `madd`? 

For the version with `mul` as the last instruction we get:
```
umulh x4, x0, x2
madd x3, x0, x3, x4
madd x1, x1, x2, x3
madd x0, x0, x2, xzr
```

For the version with the last `madd` as the last instruction, we get this:
```
umulh x4, x0, x2
madd x3, x0, x3, x4
madd x0, x0, x2, xzr
madd x1, x1, x2, x3
```

Reply:
Yes, I meant precisely your last experiment and the result is disappointing (the first move is unnecessary at the very least). Anyway, let's go with the current version, i.e. `MUL` in the last position (which still makes it possible to use the dedicated forwarding path between the pair of `MADD`s and to dual-issue), and I (or someone else) can check later what the real performance impact is and if there is some way to avoid the moves.

Reply:
@cfallin Just a quick note - now that the announcement is [public](https://community.arm.com/developer/ip-products/processors/b/processors-ip-blog/posts/first-armv9-cpu-cores), I can actually say that not only current generation, but also some of the next generation cores (i.e. Cortex-A510) are also going to be in-order.